tmc-langs-framework/src/main/java/fi/helsinki/cs/tmc/langs/domain/Filer.java

```diff
@@ -40,100 +39,80 @@ public Filer setLanguagePlugin(LanguagePlugin languagePlugin) {
-            if (file.getFileName().toString().contains(item)) {
-                return true;
-            }
+        
+        if (data.isEmpty()) {
+            logger.info("skipped file as empty while copying from: {} to:{}", source, destination);
+        } else {
+            Files.createDirectories(destination.getParent());
+            Files.write(destination, data, StandardCharsets.UTF_8, StandardOpenOption.CREATE);
         }
-        return false;
+    }
+    
+    public List<String> prepareFile(List<String> data, String fileType) {
+        List<MetaSyntax> syntaxes = MetaSyntaxGenerator.listSyntaxes(fileType);
+        for (MetaSyntax metaSyntax : syntaxes) {
+            data = filterData(data, metaSyntax);
+        }
+        return data;
+    }
+    List<String> filterData(List<String> data, MetaSyntax m) {
+        return data; // Subclasses override this method
     }
 
-    /**
-     * Copies and filters the file based on given configuration.
-     *
-     * <p>TODO: refactor, I don't like this.
-     */
-    public void maybeCopyAndFilterFile(Path file, Path repoPath, Path exercisePath) {
-//        System.out.println("file:         " + file);
-//        System.out.println("repoPath:     " + repoPath);
-//        System.out.println("exPath:       " + exercisePath);
-        Path relativePath = file.subpath(repoPath.getNameCount(), file.getNameCount());
-//        System.out.println("relativePath: " + relativePath);
-//        System.out.println("");
-        logger.info("Looking into file: {} ", file);
-        try {
-            if (skipFile(file)) {
-                logger.info("Skipping file: {} ", file);
-                return;
-            }
-            Path toFile = toPath.resolve(relativePath);
-            if (justCopy(file)) {
-                Files.createDirectories(toFile.getParent());
-                Files.copy(file, toFile, StandardCopyOption.REPLACE_EXISTING);
-                logger.info("Just copying file from: {} to:{}", file, toFile);
-            } else {
-                List<String> data = readFile(file);
-                List<String> output = prepareFile(data);
-                if (!output.isEmpty()) {
-                    Files.createDirectories(toFile.getParent());
-                    Files.write(toFile, output, StandardCharsets.UTF_8, StandardOpenOption.CREATE);
-                    logger.info("Filtered file while copying from: {} to:{}", file, toFile);
-                } else {
-                    logger.info("skipped file as empty while copying from: {} to:{}", file, toFile);
-                }
+    protected final boolean isTextFile(Path file) throws IOException {
+        List<String> skipList = Arrays.asList(new String[] {"class", "jar"});
+        if (file.toFile().isFile() && skipList.contains(getFileExtension(file))) {
+            return false;
+        }
+        // If >90% of characters in file are sourcecode-text-like, interpret as a text file
```

Comment:
Shouldn't that still be primarily fixed by having a per-`MetaSyntax` whitelist of file type suffixes rather than with a complex-yet-fragile heuristic? 

The only reason I see for keeping this is if we'd take a "supported-by-default" stance on file types, the usefulness of which I find dubious.